Sources/MongoSwift/SDAM.swift

```diff
@@ -19,28 +37,72 @@ public struct ServerAddress: Equatable {
             }
             return String(cString: baseAddress.assumingMemoryBound(to: CChar.self))
         }
+        self.type = .hostname
         self.port = hostData.port
     }
 
     /// Initializes a ServerAddress, using the default localhost:27017 if a host/port is not provided.
     internal init(_ hostAndPort: String = "localhost:27017") throws {
-        let parts = hostAndPort.split(separator: ":")
-        self.host = String(parts[0])
-        guard let port = UInt16(parts[1]) else {
-            throw MongoError.InternalError(message: "couldn't parse address from \(hostAndPort)")
+        guard !hostAndPort.contains("?") else {
+            throw MongoError.InvalidArgumentError(message: "\(hostAndPort) contains invalid characters")
+        }
+
+        var host: String
+        var port: UInt16
+
+        // Check if host is an IPv6 literal.
+        if hostAndPort.first == "[" {
+            let ipLiteralRegex = try NSRegularExpression(pattern: #"^\[(.*)\](?::([0-9]+))?$"#)
+            guard
+                let match = ipLiteralRegex.firstMatch(
+                    in: hostAndPort,
+                    range: NSRange(location: 0, length: hostAndPort.utf16.count)
+                ),
+                let hostRange = Range(match.range(at: 1), in: hostAndPort)
+            else {
+                throw MongoError.InvalidArgumentError(message: "couldn't parse address from \(hostAndPort)")
+            }
+            host = String(hostAndPort[hostRange])
+            if let portRange = Range(match.range(at: 2), in: hostAndPort) {
+                port = try ServerAddress.verifyPort(portOptional: UInt16(hostAndPort[portRange]))
+            } else {
+                port = 27017
+            }
+            self.type = .ipLiteral
+        } else {
+            let parts = hostAndPort.components(separatedBy: ":")
+            host = String(parts[0])
+            guard parts.count <= 2 else {
+                throw MongoError.InvalidArgumentError(
+                    message: "expected only a single port delimiter ':' in \(hostAndPort)"
+                )
+            }
+            if parts.count > 1 {
+                port = try ServerAddress.verifyPort(portOptional: UInt16(parts[1]))
```

Comment:
instead of having this function just verify the port, let's have it do both the parsing and verification of the port. That way the `UInt16` portion won't have to be repeated. We could then call it `parsePort(from: String)` or something along those lines.

Reply:
Had to use Substring